Firestore/core/src/firebase/firestore/local/memory_index_manager.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_INDEX_MANAGER_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_INDEX_MANAGER_H_
+
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/local/index_manager.h"
+#include "Firestore/core/src/firebase/firestore/model/resource_path.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/**
+ * Internal implementation of the collection-parent index. Also used for
+ * in-memory caching by LevelDbIndexManager and initial index population during
+ * schema migration.
+ */
+class MemoryCollectionParentIndex {
+ public:
+  // Returns false if the entry already existed.
+  bool Add(const model::ResourcePath &collection_path);
+
+  std::vector<model::ResourcePath> GetEntries(
+      const std::string &collection_id) const;
+
+ private:
+  std::map<std::string, std::set<model::ResourcePath>> index_;
```

Comment:
Note that `std::map` and `std::set` are less efficient than their `unordered_` counterparts, so they should only be used when sorted order matters. It looks like the map can be an `unordered_map`? Not sure about the `set` -- it would affect whether the `vector` returned by `GetCollectionParents` is sorted.

Reply:
More concretely: `std::map` is like a java `TreeMap` and `std::unordered_map` is like a `HashMap`.

Reply:
Okay, done.  I kept `set<ResourcePath>` as is for now as we don't have a hash defined for paths and I'd rather not tackle that in this PR.  But I can if necessary.